runner/check.go

```diff
@@ -0,0 +1,120 @@
+package runner
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os/exec"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// Check is a basic structure that describes DC/OS check.
+type Check struct {
+	// Cmd is a path to executable (script or binary) and arguments
+	Cmd []string `json:"cmd"`
+
+	// Description provides a basic check description.
+	Description string `json:"description"`
+
+	// Timeout is defines a custom script timeout.
+	Timeout string `json:"timeout"`
+
+	// Roles is a list of DC/OS roles (e.g. master, agent). Node must be one of roles
+	// to execute a check.
+	Roles []string `json:"roles"`
+}
+
+// Run executes the given check.
+func (c *Check) Run(ctx context.Context, role string) ([]byte, error) {
+	var (
+		// define an empty cancel function
+		cancel context.CancelFunc = func() {}
+
+		arg []string
+	)
+
+	if !c.verifyRole(role) {
+		return nil, errors.Errorf("check can be executed on a node with the following roles %s. Current role %s", c.Roles, role)
+	}
+
+	if len(c.Cmd) == 0 {
+		return nil, errors.New("unable to execute a command with empty Cmd field")
+	}
+
+	if ctx == nil {
+		// default to 10 seconds timeout.
+		ctx, cancel = context.WithTimeout(context.Background(), time.Second*10)
+	}
+
+	if c.Timeout != "" {
+		timeout, err := time.ParseDuration(c.Timeout)
+		if err != nil {
+			return nil, errors.Wrapf(err, "invalid timeout string %s", timeout)
+		}
+
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+	}
+	defer cancel()
+	if len(c.Cmd) > 1 {
+		arg = c.Cmd[1:]
+	}
+
+	cmd := exec.CommandContext(ctx, c.Cmd[0], arg...)
+
+	// create stdout/stderr pipes for combined output.
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to open stdout pipe")
+	}
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to open stderr pipe")
+	}
+
+	// start command execution
+	if err := cmd.Start(); err != nil {
+		return nil, errors.Wrapf(err, "unable to run command %s", c.Cmd)
+	}
+
+	// buf will contain a combined output.
+	buf := new(bytes.Buffer)
+
+	// readErr will be used to pass back error by io.Copy
+	readErr := make(chan error, 1)
+
+	// execute the reader in the background. This way we will not block command termination
+	// if timeout reached.
+	go func() {
+		mr := io.MultiReader(stdout, stderr)
```

Comment:
The golang docs say that the `MultiReader` reads from the readers sequentially.  So it seems like your `buf` will first contain all of the stdout lines, and then all of the stderr lines.  That seems like it could be really confusing since the chronology would be misrepresented if a program interleaves stdout and stderr (which is totally possible).